common/utils/utils.go

```diff
@@ -78,3 +81,88 @@ func ReadGenesis(genesisPath string) (*core.Genesis, error) {
 	}
 	return genesis, file.Close()
 }
+
+// OverrideConfigWithEnv recursively overrides config values with environment variables
+func OverrideConfigWithEnv(cfg interface{}, prefix string) error {
+	v := reflect.ValueOf(cfg)
+	if v.Kind() != reflect.Ptr || v.IsNil() {
+		return nil
+	}
+	v = v.Elem()
+
+	t := v.Type()
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+		fieldValue := v.Field(i)
+
+		if !fieldValue.CanSet() {
+			continue
+		}
+
+		tag := field.Tag.Get("json")
+		if tag == "" {
+			tag = strings.ToLower(field.Name)
+		}
+
+		envKey := prefix + "_" + strings.ToUpper(tag)
+
+		switch fieldValue.Kind() {
+		case reflect.Ptr:
+			if !fieldValue.IsNil() {
+				err := OverrideConfigWithEnv(fieldValue.Interface(), envKey)
+				if err != nil {
+					return err
+				}
+			}
+		case reflect.Struct:
+			err := OverrideConfigWithEnv(fieldValue.Addr().Interface(), envKey)
+			if err != nil {
+				return err
+			}
+		default:
+			if envValue, exists := os.LookupEnv(envKey); exists {
```

Comment:
This log will make debugging much easier.

```suggestion
			if envValue, exists := os.LookupEnv(envKey); exists {
				log.Info("Overriding config with env var", "key", envKey)
```

Reply:
A side effect: it will also print the private key, could check the use cases (if it's acceptable).

Reply:
That's why I suggest to only print the key and not the value.

Reply:
oh I see.